### DISCUSSION TITLE
odd register spills in function prologue

### DISCUSSION THREAD

**Discussion by mindbeast** [ispc/ispc#2173] [Q&A] [answered]
I've hit a few cases where varying pod structs end up spilling to memory for what seems like unnecessary reasons in the function prologue.

Here's a [toy example](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1B4FAB2Skl9ZATwDKjdAGFUtAK4sGEgGylHAGTwGTAA5DwAjTGIQSQB2UlNUBUI7Bhd3Tx8EpJSBQOCwlkjouMtMa1sBIQImYgJ0jy8uXytMG1Tq2oJ80IiomPiFGrqGzObLYe6g3qL%2BuIBKS1Q3YmR2DiHiNxsAagBBKhpgnYBSWIAhE409ndud9GXw%2Bh3iE4BWc4Bmd4ARd6/fidPpdrnd7o9ngR/t83n9gVc9mc4SDEddNtsCDtlKgggRPuhThcEWCHm4nph7kwatDAfDrkigSiEScAEz4zBHCkAFT2ACUAOLYLkAfQA6gBJH5cgASO0kzLZWE5OwAsnsABrC5QAeXFIS5Qh2LLe3mZaIIW124oYpjcmLOKLupPJOyCwtU/zVmp1eoNgI%2BPIFQrFkpltMdt2dzzdAE9PRqtbr9UJ/edA4KRRKpdLw8TIxCKW6AF7x71Jv2w/7p4NZsOVuk3fNkyHEQwKfjEFjC14fGF/Xup6uZ0M5%2BsR8HNikWtsdrtQgeVgN8jMh7O50G3XE7RK4hSM5mxZFmvbo3bau22%2B1EjcTl3LAju0uJ30pxdp5c1kfrxu35734Vxh8XrPsmg4fsOa5jnmv4Uv%2BJZAQmPqgW%2BQ6rnWR43luO6CHuDYMnh5pUngyCugwtDTFiOKCPiOxMKYpi0DGwpMIc0wQGgDBDPsrHHKy3gsZypA7BxXHYriNF8dhBDzIS46UeJBL3vu9JsqYrbACwTA7G4DDEK4tDQTsHY7BAOl4LOpGYngpyfD8OwaIyro2U4OzfMCrqspcLLnDJDqGSSBY7AoHg2XZGgAHQOXhqnqZp2m6fp/l3MZpkMOZJAsJZOzaKF9mOTlQIuW55zZZ5nm%2BdeP5gtVwWZWVtm0TxmDhT25x4Km2i/DsABU25UQQ4XoFSTD/J1Y5gkltwMje1X3oNw3/O1sK5QJ0zhfObVdZ5QUeMpVVIhNM3EJgBArAwOxKXhh7HpgqiJHUOwAG44gS06cbO7ECFxZkWdal6nCy3ikZeQk/RlOzngQ/18fel7zAecmPbUMZBMA3HKm97YZXtCrssqQjKOK/j%2BDsXDHqynzmTsBNE/4uNqUwGlaTpem0AZM0pWDnZZdZQJhY5vOfEVAtld5FVMjFjNxSziUzWCnNpRZW4FQ1UUlSrwvuZ13nlbJk3VZjs4tYtHVbQ1QSXutrbvRl3Ym2%2BY39ucamoMArYsNaWAelBct3NNVVgobGXrfbdl88DdpWzOtsbUtTsu27rCe7d34TddKmfOUSiGUHnbGx8DlvoXYfmzake512rXF/81cfAn7vJ97GFVRX%2BfnLX5xk8t4cW%2BX1tY52dsF6mXfx3picewwXup3crdV6mLJm3ZvcDRXQ/twvqb10nU8pz7Lf90brWjzXS8R6vh%2B28fqYd9vk/T/vgeX3n19vqPuUr1HNuD6/Tsn3X48G67ybntJ%2B0cX7/H/ucRe3dS6WzXr/f4MCx6uyAQ/ZuYDv4sDbsg0%2BsDl5lwvuAyuSCb5b0ATvdBoDZ7P2wa1XBHx3490IV/AeJCPgMM7uQ1BlC94YJocQnBm98Hn1YbOdenDOF30bjPW4rcNrFw/iwteCjuETxkY/ARWCQ6MLPp/FRkC1FoL4dQuRtCdHQL0co2hwoNpSIoffEx0VM5T3MseMEAB6DxOwuS0NomzHYBAEAUikgocKhlYaR1CUoy2oScYzQpgzJmOwGDLASmzQyCt0rcy3ILfm7lBYuU/nEgposfJ619rcJGxAUZGHktRAkUlTHVSkvNakw8RGfzdN7Tab5pHANkS0/qbSRq6M6Sw2Mod/j9KoQ2aqdxWlDXaZY8ZltixTIATwxxICCIBzuNU2paMxINIunaFa9FGLMSahACuQkpLw12fM24kSBr/h6XHaZDiNElzsnNJZoyN6aPmS88K/5AK9JQeogZIi/kLTGfw4FF5I5wQ2c7L50KfmnIGv80hQLCR/HpOnPYHBFi0E4G8XgXgOBaFIKgTgJhzBBWWKsCkFMeCkAIJoEliwADWMQWThWaJ8DQkhRUAA5YgaFiG8Lgbw3j6E4JISlXLaWcF4AoEAGgOVcsWHAWASA0AsFMHQKI5BKCGuNfQaIDLkAAFoLQ6R5XwOgBAogaogOEFV4QgjI04Oyw1bBBDajIjGFVWBNJGHENS3g%2BBjrtEepgDV0a/CqDaHadY7LcTlBVeRcIrYakuCwCqi0eAWB%2BpJXwAwwAFAADU8CYAAO7alMIwctMhBAiDEOwKQ7b5BKDUCq3QXB9CGGMGYCwuaNWQEWKgUwlROKcFtdqT4OxbX2tqMAE6fMmCPVUCyW1eBvCSFUJIVdtrkC2j5goHlqh1XlDaPOhwU9RhND8FPHohRih6ESMkedL7v05HnR%2Bvo0Rh2tHaFUSY/6wP3ogwwTodRgOzFAxMLo0HUOIemJ%2B/oXBFgKGZWsPQFpMAZorWSjgFLSBUppXSjgNqTIOoYDymSEAbVCVwIQEgANPjDp2C4I1JriDcdw7wTl0b5iLGCUwLA0QIC8pAJ8MV4U3hioAJyfBZKp7wbJmiytUwqjgSqqMqto%2BqzV2rxOkD1YgEALyzWfQE1avQmB8BEBQ/wDtohxA9o832lQ6hk3EdI6QBtrZTDltJeS5VybaOQ3%2BqgKgrpx0Ma2Exlj/HLVRGE/MUTOr5NcFU%2BFFkcRtMqbU7EbwGnvDxHI0Z6jvBTOWHM2JrQEmDMsmizRtVFnWuLATcQZI9hJBAA) in godbolt that shows one instance of this. 

You can flip the SPILL define between 0 and 1. When spill is 1, the entire Affine struct is spilled to memory and then reloaded, when spill is 0 it's loaded directly into registers with no spilling. Seems pretty odd to me. Any idea why there would be any difference between the two codepaths? Salient bit:

```
struct Affine {
    double r[3][3];
    double t[3];
};
```

**No-spill path (manually unrolled)**

```
    varying Affine transform;
    transform.r[0][0] = input.transform_r[0][0][programIndex];
    transform.r[0][1] = input.transform_r[0][1][programIndex];
    transform.r[0][2] = input.transform_r[0][2][programIndex];
    transform.r[1][0] = input.transform_r[1][0][programIndex];
    transform.r[1][1] = input.transform_r[1][1][programIndex];
    transform.r[1][2] = input.transform_r[1][2][programIndex];
    transform.r[2][0] = input.transform_r[2][0][programIndex];
    transform.r[2][1] = input.transform_r[2][1][programIndex];
    transform.r[2][2] = input.transform_r[2][2][programIndex];
    transform.t[0] = input.transform_t[0][programIndex];
    transform.t[1] = input.transform_t[1][programIndex];
    transform.t[2] = input.transform_t[2][programIndex];

```

**Spilling path**

```
    varying Affine transform;
#pragma unroll
    for (uniform int i = 0; i < 3; i++) {
#pragma unroll
        for (uniform int j = 0; j < 3; j++) {
            transform.r[i][j] = input.transform_r[i][j][programIndex];
        }
        transform.t[i] = input.transform_t[i][programIndex];
    }

```

**Comment by mindbeast** [comment]
I see the same behavior no matter how small the struct is. The current struct in the example is pretty big, 12 doubles per program index, but there is plenty of available register space to place as, as evidenced by the non-spilling path.

**Comment by dbabokin** [comment, marked as answer]
Not sure what's going on here, but it's definitely a bug. I've created issue #2174 for it.
